In [8]:
import pandas as pd 
import numpy as np


import warnings
warnings.simplefilter('ignore')

In [9]:
df = pd.read_excel(r"E:\Data Science by SRK\Github\Machine_learning\Regression_projects\Property_price_prediction\property_price_prediction_data.xlsx")

In [10]:
df

size_sq_ft  propertyType  bedrooms      price  AP_dist_km  \
0        5.993961             1         1   9.105091    3.124178   
1        6.957497             0         2   9.903538    3.109144   
2        7.719130             1         2  10.239996    3.162109   
3        7.208600             1         2  10.239996    3.164682   
4        6.111467             0         2   9.433564    3.160942   
...           ...           ...       ...        ...         ...   
17700    5.602119             1         1   9.615872    2.693788   
17701    6.908755             1         3  10.545368    2.693810   
17702    7.244942             1         1  10.757924    2.625941   
17703    6.803505             1         1   8.987322    2.660821   
17704    6.908755             1         3  10.545368    2.711261   

       Aiims_dist_km  NDRLW_dist_km  suburb_Delhi Central  suburb_Delhi East  \
0           2.494794       2.481289                 False               True   
1           2.343637       3.228757                 False               True   
2           2.551746       2.736483                 False               True   
3           2.563618       2.720278                 False               True   
4           2.448552       3.450400                 False               True   
...              ...            ...                   ...                ...   
17700       2.237752       1.439911                 False              False   
17701       2.237903       1.440214                 False              False   
17702       2.220053       1.827009                 False              False   
17703       2.269011       1.753226                 False              False   
17704       2.265247       1.400796                 False              False   

       suburb_Delhi North  suburb_Delhi South  suburb_Delhi West  \
0                   False               False              False   
1                   False               False              False   
2                   False               False              False   
3                   False               False              False   
4                   False               False              False   
...                   ...                 ...                ...   
17700               False               False               True   
17701               False               False               True   
17702               False               False               True   
17703               False               False               True   
17704               False               False               True   

       suburb_Dwarka  suburb_North Delhi  suburb_North West Delhi  \
0              False               False                    False   
1              False               False                    False   
2              False               False                    False   
3              False               False                    False   
4              False               False                    False   
...              ...                 ...                      ...   
17700          False               False                    False   
17701          False               False                    False   
17702          False               False                    False   
17703          False               False                    False   
17704          False               False                    False   

       suburb_Other  suburb_Rohini  suburb_South West Delhi  suburb_West Delhi  
0             False          False                    False              False  
1             False          False                    False              False  
2             False          False                    False              False  
3             False          False                    False              False  
4             False          False                    False              False  
...             ...            ...                      ...                ...  
17700         Fa

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17705 entries, 0 to 17704
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   size_sq_ft               17705 non-null  float64
 1   propertyType             17705 non-null  int64  
 2   bedrooms                 17705 non-null  int64  
 3   price                    17705 non-null  float64
 4   AP_dist_km               17705 non-null  float64
 5   Aiims_dist_km            17705 non-null  float64
 6   NDRLW_dist_km            17705 non-null  float64
 7   suburb_Delhi Central     17705 non-null  bool   
 8   suburb_Delhi East        17705 non-null  bool   
 9   suburb_Delhi North       17705 non-null  bool   
 10  suburb_Delhi South       17705 non-null  bool   
 11  suburb_Delhi West        17705 non-null  bool   
 12  suburb_Dwarka            17705 non-null  bool   
 13  suburb_North Delhi       17705 non-null  bool   
 14  suburb_North West Delh

# X and y

In [12]:
X = df.drop(columns = ['price'])
y = df['price']

# Train_Test_Split

In [13]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 5)

# Modelling

In [14]:
# ALL EVALUATION IN ONE CELL

from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.model_selection import cross_val_score


from sklearn.linear_model import LinearRegression

model = LinearRegression()

model.fit(X_train, y_train)


print('coefficient : ', model.coef_)
print('itercept : ', model.intercept_)
print("===========================")


# Prediction RMSE

ypred_train = model.predict(X_train)
ypred_test = model.predict(X_test)


# EVALUATION

print("Train_RMSE : ", np.sqrt(mean_squared_error(y_train, ypred_train)))
print("cv_score_RMSE", np.sqrt(abs(cross_val_score(model, X_train, y_train, cv = 5, scoring = 'neg_mean_squared_error')).mean()))
print("Test_RMSE : ", np.sqrt(mean_squared_error(y_test, ypred_test)))


print("=================================================")


# Prediction R2

ypred_train = model.predict(X_train)
ypred_test = model.predict(X_test)

print("Train_R2 : ", r2_score(y_train, ypred_train))
print("CV_Score_R2", cross_val_score(model, X_train, y_train, cv = 5, scoring = 'r2').mean())
print("Test_R2 : ", r2_score(y_test, ypred_test))

coefficient :  [ 7.08023793e-01  6.04774207e-03  2.04250574e-01  1.58381122e-03
 -3.25329519e-01 -1.33211299e-01 -7.72377498e-02 -2.73985356e-01
  1.46028664e-01 -2.96109306e-02  1.31953678e-01 -2.97766755e-02
  3.83964047e-02  1.14504328e-01 -6.78106095e-02  1.12620617e-01
 -6.48216967e-02 -2.60673239e-04]
itercept :  6.011973798257787
Train_RMSE :  0.3531550236430932
cv_score_RMSE 0.35387813668716916
Test_RMSE :  0.3655763210546383
Train_R2 :  0.7751593739466244
CV_Score_R2 0.774021623350008
Test_R2 :  0.751888342689825


# Variable significance check

In [15]:
from sklearn.model_selection import train_test_split
import statsmodels.formula.api as smf

# Splitting the data
df_train, df_test = train_test_split(df, train_size=0.8, random_state=5)

# Use formula-style notation with column names
formula = "price ~ size_sq_ft + bedrooms + AP_dist_km + Aiims_dist_km + NDRLW_dist_km"  # Include relevant columns

# Fitting the model
model1 = smf.ols(formula, data=df_train).fit()
print(model1.summary())


                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.762
Model:                            OLS   Adj. R-squared:                  0.762
Method:                 Least Squares   F-statistic:                     9072.
Date:                Thu, 17 Apr 2025   Prob (F-statistic):               0.00
Time:                        12:12:32   Log-Likelihood:                -5754.4
No. Observations:               14164   AIC:                         1.152e+04
Df Residuals:                   14158   BIC:                         1.157e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         5.9330      0.065     91.544

**Observation**
 The model explains 76.2% of the variance in property prices (R² = 0.762), with size_sq_ft, bedrooms, and distances to landmarks being significant predictors (p < 0.05 for all variables).

In [16]:
# Save the model

import joblib

# Save model
joblib.dump(model, 'property_price_model.joblib')


['property_price_model.joblib']